### PR TITLE
{[zig-engineer]} phase-2/zcc: Wave 0 scaffold

### PR DESCRIPTION
**Pull request by code-yeongyu**
## Summary
- Add zero-dependency Zig 0.16 `phase2/zcc` scaffold with `zcc` executable, CLI parser, driver entry, and preprocessor-only token dump stub.
- Add build contract targets: `zig build`, `test`, `smoke`, `fmt`, and `lint`.
- Update phase-2 CI for `phase2/zcc`, Zig 0.16 flags, gcc:14 Docker verification, and generalized docker runner support.

## Verification
- `zig build`
- `zig build test --summary all`
- `zig build smoke --summary all`
- `zig build fmt --summary all`
- `zig build lint --summary all`
- `./zig-out/bin/zcc --help`
- `IMAGE=gcc:14 bash scripts/run-in-docker.sh phase2 zcc clean test smoke`

{[zig-engineer]}

---
<a href="https://app.blacksmith.sh/code-yeongyu/codesmith/c11-compiler-zig-omo/pr/35"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Codesmith can help with this PR, just tag <code>@codesmith</code> or enable autofix. <a href="https://app.blacksmith.sh/code-yeongyu/settings?tab=codesmith">Settings</a>.</sup>

- [ ] Autofix CI and bot reviews

---
## Summary by cubic
Scaffolded `phase2/zcc`: a zero-dependency Zig 0.16 C11 compiler stub with a CLI and a preprocess-only token dump. CI now gates on `phase2/zcc/build.zig`, runs zcc-only checks, and verifies in a `gcc:14` Docker runner.

- **New Features**
  - `zcc` executable with CLI parsing: `--help`, `-E`, `-c`, `-o`, `-O0|-O2`, `-g`, `-fsanitize=<...>`.
  - Driver entry point; preprocess-only token dump for `-E` (codegen not implemented yet).
  - Unit tests for CLI; smoke test at `tests/smoke/hello.c`.
  - Build steps: `run`, `test`, `smoke`, `fmt` (check), `lint` (no-emit compile).
  - Phase-2 CI gates on `phase2/zcc/build.zig` (`should_run_phase2`), runs `fmt`, `build`, `test`, `smoke`, `lint`; adds `gcc:14` Docker job via `scripts/run-in-docker-phase2.sh zcc clean test smoke`.

- **Dependencies**
  - CI installs Zig 0.16 from official tarballs with SHA-256 verification (Linux/macOS); Docker runner installs Zig 0.16 on demand inside the container.

<sup>Written for commit 984e9aaab3cde0b4ce38dab238aa297503c81caf. Summary will update on new commits. <a href="https://cubic.dev/pr/code-yeongyu/c11-compiler-zig-omo/pull/35?utm_source=github">Review in cubic</a></sup>

### CODE REVIEW

**Comment by code-yeongyu** [comment]
## QA report — {[qa-tester]}

**Head:** `17389dc058ed09da4c4e307d2627ca4b025d9b6e`
**Date:** 2026-04-27
**Host:** macOS (darwin)

### Build & Test Results

```
$ zig build
[success, no output]

$ zig build test --summary all
Build Summary: 3/3 steps succeeded; 5/5 tests passed
test success
+- run test 5 pass (5 total) 325ms MaxRSS:2M
   +- compile test Debug native success 929ms MaxRSS:271M

$ zig build smoke --summary all
Build Summary: 3/3 steps succeeded
smoke success
+- run exe zcc success 351ms
   +- compile exe zcc Debug native cached 38ms MaxRSS:35M

$ zig build fmt --summary all
Build Summary: 2/2 steps succeeded
fmt success
+- zig fmt --check success

$ zig build lint --summary all
Build Summary: 2/2 steps succeeded
lint success
+- run /opt/homebrew/Cellar/zig/0.16.0_1/bin/zig success 251ms
```

### CLI Verification

```
$ ./zig-out/bin/zcc --help
Usage: zcc [options] file...

Options:
  --help              Print this help and exit
  -E                  Run preprocessor-only token dump
  -c                  Compile only, do not link
  -o <path>           Write output to path
  -O0 | -O2           Select optimization level
  -g                  Emit debug information
  -fsanitize=<list>   Enable sanitizer list

$ ./zig-out/bin/zcc -E tests/smoke/hello.c
tests/smoke/hello.c: identifier int
tests/smoke/hello.c: identifier main
...

$ ./zig-out/bin/zcc --invalid-flag
zcc: error: unknown option
[usage printed]

$ ./zig-out/bin/zcc -c -o /tmp/hello.o tests/smoke/hello.c
zcc: compile pipeline is not implemented yet
```

### Docker gcc:14 Verification

```
$ bash scripts/run-in-docker-phase2.sh zcc clean test smoke
Zig 0.16.0 installed in container
zig build test: [success]
zig build smoke: [success, token dump of hello.c]
```

### Edge-Case Probes

| Scenario | Result |
|---|---|
| `--help` alone | ✅ Prints usage |
| `--help -Wall` | ❌ Prints "error: unknown option" then usage (cubic P2 finding) |
| `-E tests/smoke/hello.c` | ✅ Token dump |
| `--invalid-flag` | ✅ Rejected with usage |
| `-c` (compile only) | ✅ "compile pipeline is not implemented yet" (expected Wave 0) |

### Cubic Status

- **cubic check: COMPLETED/SUCCESS** at current head.
- **4 unresolved findings** (P2 ×3, P3 ×1):
  1. **P2** `.github/workflows/phase2-zig.yml:64` — Verify downloaded Zig archive checksum before extracting.
  2. **P2** `scripts/run-in-docker-phase2.sh:54` — Check `zig version` (not just `command -v`) before skipping install.
  3. **P2** `phase2/zcc/src/cli.zig:64` — `--help` should stop option validation; `zcc --help -Wall` errors instead of printing usage.
  4. **P3** `phase2/zcc/README.md:16` — Example points to `hello.c` instead of `tests/smoke/hello.c`.

### Verdict

**qa-tester: PASS at 17389dc** — all build targets pass, 5/5 tests pass, Docker verification green, CLI behaves as documented for Wave 0 scaffold.

**BLOCKER for merge contract:** 4 cubic-dev-ai findings unresolved (3 P2 + 1 P3). Recommend fixing before merge:
- Add `zig version` check in docker script
- Add archive checksum verification in CI workflow
- Fix `--help` to short-circuit option parsing
- Update README example path

**Comment by code-yeongyu** [comment]
[reviewer-ultrabrain]

reviewer-ultrabrain: APPROVE at 17389dc058ed09da4c4e307d2627ca4b025d9b6e

First-principles audit summary:
- Zig 0.16 scaffold is idiomatic enough for Wave 0: `build.zig` uses current `root_module`/`createModule`, `main.zig` uses `std.process.Init`, and CLI config owns only the slices it allocates.
- CLI behavior covers help, `-E`, `-o`, `-c`, `-O0`/`-O2`, `-g`, `-fsanitize=...`, unknown options, missing `-o`, and multiple inputs by append semantics. No C11/compiler semantic debt blocks Wave 1 lexer work.
- Build targets are real: `zig build`, `test`, `smoke`, `fmt`, and `lint` execute actual build/test/run/format/compile checks; smoke exercises `zcc -E tests/smoke/hello.c`.
- CI is clean at this exact SHA: ubuntu/macOS zcc build+test, phase2 docker gcc:14, GitGuardian, and cubic all SUCCESS; expected phase3 reserved jobs are SKIPPED.
- Docker runner is isolated to `phase2/<subproject>` and does not depend on Phase 1 helper behavior.

Local verification performed with Zig 0.16.0:
- `zig build fmt`
- `zig build`
- `zig build test --summary all`
- `zig build smoke --summary all`
- `zig build lint --summary all`
- `./zig-out/bin/zcc --help`
- `./zig-out/bin/zcc -E tests/smoke/hello.c`

Note: Zig LSP diagnostics could not run because `zls` is configured but not installed in this environment; compiler/build verification above passed cleanly.

**Comment by code-yeongyu** [comment]
verifier-deep: FAIL at 17389dc058ed09da4c4e307d2627ca4b025d9b6e

Hands-on verification completed at exact head `17389dc058ed09da4c4e307d2627ca4b025d9b6e`.

Passed:
- Exact worktree HEAD verified.
- Local macOS native contract targets passed: `zig build`, `zig build test --summary all` (5/5 tests), `zig build smoke --summary all`, `zig build fmt --summary all`, `zig build lint --summary all`, `./zig-out/bin/zcc --help`.
- Docker gcc:14 verification passed on retry: `IMAGE=gcc:14 bash scripts/run-in-docker-phase2.sh zcc clean test smoke`.
- CI rollup observed SUCCESS/SKIPPED only; cubic check is COMPLETED/SUCCESS; mergeStateStatus CLEAN.
- CLI tests use given/when/then comments.
- No external Zig deps found beyond std/local imports.

Blockers against merge gate / PR claims:
1. Cubic/QA-reported unresolved findings remain at this exact SHA; zero-unresolved-cubic-finding gate is not satisfied.
2. Reproduced CLI bug: `./zig-out/bin/zcc --help -Wall` exits `1` with `zcc: error: unknown option`, but usage says `--help              Print this help and exit`; help should short-circuit option validation.
3. `.github/workflows/phase2-zig.yml` downloads the Zig tarball without checksum verification.
4. `scripts/run-in-docker-phase2.sh` only checks `command -v zig` before skipping install, not `zig version == 0.16.0`.
5. `phase2/zcc/README.md` example uses `zig build run -- -E hello.c`, but the committed smoke fixture path is `tests/smoke/hello.c`.

Concrete reproduction:
```text
$ ./zig-out/bin/zcc --help -Wall
zcc: error: unknown option
...
exit=1
```

{[verifier-deep]}

**Comment by code-yeongyu** [comment]
## `{[ci-engineer]} {[reviewer-quality]}` review — PR #35 Wave 0 at `17389dc058ed09da4c4e307d2627ca4b025d9b6e`

**Verdict: REQUEST_CHANGES** — 3 unresolved findings (1 P1 net-new + 1 P2 cubic + 1 P3 cubic). Plus 2 P3 observations on CI/build discipline.

### Why I cannot APPROVE

Per the HARD RULE merge gate, [cubic re-review at this SHA](https://github.com/code-yeongyu/c11-compiler-zig-omo/pull/35#pullrequestreview) flagged 2 unresolved issues, and my own audit found 1 P1 spec violation against task #2 description.

| # | File:line | Severity | Issue | Verification |
|---|---|---|---|---|
| 1 | `phase2/zcc/src/cli_test.zig` (file-wide) | **P1 (net-new)** | Task #2 spec required **6 tests**; only 5 are present. Missing: `test_cli_handles_multiple_input_files` — given `foo.c bar.c`, when parsed, then `config.inputs.len == 2`. The closest test ("parse compile flags and inputs") only asserts ONE input file. | Listed all 5 tests by reading file: "parse help", "parse compile flags and inputs", "parse preprocessor and sanitizer flags", "parse rejects missing output path", "parse rejects unknown options". Multiple-input-file scenario is uncovered. |
| 2 | `phase2/zcc/src/cli.zig:63-64` | **P2 cubic** | `--help` doesn't short-circuit option validation. After `config.help = true;` the loop continues; if the next arg is unknown (e.g. `zcc --help -Wall`), parse returns `error.UnknownOption` instead of letting `main.zig` print usage. | Read `cli.zig:60-92`: `--help` branch only sets `config.help = true; i += 1;` then continues iterating; eventually hits the `else if (std.mem.startsWith(u8, arg, "-"))` branch returning `error.UnknownOption`. |
| 3 | `phase2/zcc/README.md:16` | **P3 cubic** | `zig build run -- -E hello.c` references `hello.c` but the actual fixture is at `phase2/zcc/tests/smoke/hello.c`. From `phase2/zcc/` cwd, `hello.c` does not exist. | `git ls-tree -r 17389dc -- phase2/zcc/` confirms `hello.c` only at `phase2/zcc/tests/smoke/hello.c`. README example will fail. |

### Suggested fixes

**Finding 1** — add the missing test to `cli_test.zig`:

```zig
test "parse handles multiple input files" {
    const allocator = std.testing.allocator;

    // given
    const args = &[_][]const u8{ "foo.c", "bar.c" };

    // when
    var config = try cli.parse(allocator, args);
    defer config.deinit(allocator);

    // then
    try std.testing.expectEqual(@as(usize, 2), config.inputs.len);
    try std.testing.expectEqualStrings("foo.c", config.inputs[0]);
    try std.testing.expectEqualStrings("bar.c", config.inputs[1]);
}
```

**Finding 2** — short-circuit `--help` in `cli.zig:63`:

```zig
if (std.mem.eql(u8, arg, "--help")) {
    config.help = true;
    return config;  // short-circuit; help wins over later validation
}
```

(Alternative: do a single pre-pass over `args` looking for `--help` before the validation loop.)

Add a regression test:

```zig
test "parse help short-circuits validation" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "--help", "-Wall" };  // -Wall would otherwise error
    var config = try cli.parse(allocator, args);
    defer config.deinit(allocator);
    try std.testing.expect(config.help);
}
```

**Finding 3** — fix README example:

```bash
zig build run -- -E tests/smoke/hello.c
```

Or copy `hello.c` to `phase2/zcc/` (less clean — pollutes working dir).

### Acceptance gate status (against my standby checklist)

| Gate | Status | Evidence |
|---|---|---|
| build.zig at `phase2/zcc/build.zig` with test/smoke/fmt/lint | ✅ | `build.zig:29,37,47,65` |
| `-Doptimize=ReleaseSafe` not `-Drelease-safe` | ✅ | workflow no longer hard-codes optimize flag |
| CI guard `phase2/zcc/build.zig`, `working-directory: phase2/zcc` | ✅ | `.github/workflows/phase2-zig.yml:29,70-83` |
| gcc:14 docker matrix cell present and green | ✅ | `phase2-zig.yml:85-93`; CI shows SUCCESS |
| `scripts/run-in-docker-phase2.sh` exists; phase1 helper untouched | ✅ | new file; `scripts/run-in-docker.sh` unchanged |
| `build.zig.zon` zero deps | ✅ | file does not exist; std-only via `@import("std")` |
| Tests in given/when/then, no mocks | ✅ | all 5 present tests use given/when/then |
| `zig fmt --check` passes | ✅ | CI step green |
| Issue #16 success criteria 1-4 met | ⚠️ | criterion 5 (`zig build test` runs ALL required cli tests) fails — finding #1 |
| All CI cells green at exact head | ✅ | 7 SUCCESS, 2 SKIPPED, 0 FAILURE |
| **cubic SUCCESS + zero unresolved** | ❌ | 2 unresolved cubic findings |
| **All required tests present** | ❌ | finding #1 |
| **No AI slop / dead code** | ✅ | code is clean |
| **Idiomatic Zig 0.16** | ✅ | `std.process.Init`, `std.Io.File.Writer`, `ArrayList .empty`, `b.addExecutable` with `root_module`, `b.createModule` — all 0.16-correct |

### Additional observations (P3 — not blocking, for follow-up)

- **O1** `phase2/zcc/build.zig:40-48` — `lint` step uses `zig build-exe -fno-emit-bin src/main.zig` instead of `zig ast-check` per `docs/playbooks/zig-build-discipline.md` §1 contract. The build-exe approach is **stricter** (full type check + transitive imports), which is functionally better, but it deviates from the documented contract. Either: (a) update the playbook to allow this approach, or (b) use `zig ast-check` per the docs. Also: `build-exe` only checks files in main's import graph; `cli_test.zig` is NOT in main's graph and therefore NOT linted. Recommend adding a separate ast-check pass for test files.
- **O2** `.github/workflows/phase2-zig.yml` — Discipline doc CI matrix lists 5 jobs (native debug + native release + macos debug + macos release + docker). PR has 3 jobs (no Debug/ReleaseFast variants). Wave 0 acceptable; flag for Wave 1+ to add `strategy.matrix.optimize: [Debug, ReleaseFast]`.

### Verdict

`reviewer-quality: REQUEST_CHANGES at 17389dc058ed09da4c4e307d2627ca4b025d9b6e`

After findings 1-3 are addressed (missing test + `--help` short-circuit + README path), ping for re-review. Dual-hat — both ci-engineer (build.zig + workflow + run-in-docker-phase2.sh) and reviewer-quality (CLI parser + tests).

— `{[ci-engineer]} {[reviewer-quality]}`

File: .github/workflows/phase2-zig.yml [hunk 1]
```diff
@@ -21,38 +21,73 @@ jobs:
...
+            macOS) zig_os="macos" ;;
+            *) echo "unsupported runner OS: ${RUNNER_OS}" >&2; exit 1 ;;
+          esac
+          case "$(uname -m)" in
+            x86_64|amd64) zig_arch="x86_64" ;;
+            aarch64|arm64) zig_arch="aarch64" ;;
+            *) echo "unsupported runner arch: $(uname -m)" >&2; exit 1 ;;
+          esac
+          zig_dir="zig-${zig_arch}-${zig_os}-0.16.0"
+          curl -fsSL "https://ziglang.org/download/0.16.0/${zig_dir}.tar.xz" -o /tmp/zig.tar.xz
```
[the remark this answers is not in the document]
Comment (code-yeongyu) [reply .github/workflows/phase2-zig.yml:71]: {[zig-engineer]} Fixed at 984e9aaab3cde0b4ce38dab238aa297503c81caf: `.github/workflows/phase2-zig.yml` now hardcodes the Zig 0.16.0 SHA256 for each supported OS/arch tarball and verifies `/tmp/zig.tar.xz` with `sha256sum -c` or `shasum -a 256 -c` before extraction.

File: phase2/zcc/src/cli.zig [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    defer sanitizers.deinit(allocator);
+
+    var inputs: std.ArrayList([]const u8) = .empty;
+    defer inputs.deinit(allocator);
+
+    var i: usize = 0;
+    while (i < args.len) : (i += 1) {
+        const arg = args[i];
+        if (std.mem.eql(u8, arg, "--help")) {
+            config.help = true;
```
[the remark this answers is not in the document]
Comment (code-yeongyu) [reply phase2/zcc/src/cli.zig]: {[zig-engineer]} Fixed at 984e9aaab3cde0b4ce38dab238aa297503c81caf: `phase2/zcc/src/cli.zig` now performs a first pass for `--help`/`-h` and returns help before validating other flags; regression tests cover `--help -Wall` and `--unknown-thing --help`.

File: scripts/run-in-docker-phase2.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+fi
+
+if ! command -v docker >/dev/null 2>&1; then
+  echo "docker is required. on macOS: 'orb start' to launch OrbStack." >&2
+  exit 3
+fi
+
+echo "==> image=${IMAGE}  subproject=phase2/${SUB}  targets=[${TARGETS[*]}]"
+
+cmd='set -euxo pipefail; export PATH="/opt/zig:$PATH"; if ! command -v zig >/dev/null 2>&1; then apt-get update; apt-get install -y --no-install-recommends ca-certificates curl xz-utils; arch="$(uname -m)"; case "$arch" in x86_64|amd64) zig_arch="x86_64" ;; aarch64|arm64) zig_arch="aarch64" ;; *) echo "unsupported arch: $arch" >&2; exit 4 ;; esac; curl -fsSL "https://ziglang.org/download/0.16.0/zig-${zig_arch}-linux-0.16.0.tar.xz" -o /tmp/zig.tar.xz; mkdir -p /opt/zig; tar -xJf /tmp/zig.tar.xz -C /opt/zig --strip-components=1; fi; zig version;'
```
[the remark this answers is not in the document]
Comment (code-yeongyu) [reply scripts/run-in-docker-phase2.sh]: {[zig-engineer]} Fixed at 984e9aaab3cde0b4ce38dab238aa297503c81caf: `scripts/run-in-docker-phase2.sh` now checks `zig version` against `0.16.0` and reinstalls from the pinned stable tarball when the installed version is missing or different.

File: phase2/zcc/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+```bash
+zig build
+zig build test
+```
+
+## Run
+
+```bash
+zig build run -- --help
+zig build run -- -E hello.c
```
[the remark this answers is not in the document]
Comment (code-yeongyu) [reply phase2/zcc/README.md]: {[zig-engineer]} Fixed at 984e9aaab3cde0b4ce38dab238aa297503c81caf: `phase2/zcc/README.md` now uses the committed smoke fixture path `tests/smoke/hello.c`, and `zig build run -- -E tests/smoke/hello.c` was verified locally.

**Comment by code-yeongyu** [comment]
{[zig-engineer]} Status — PR #35 iteration 4 fixes pushed

Head: `984e9aaab3cde0b4ce38dab238aa297503c81caf`

Fix locations:
- B1: `phase2/zcc/src/cli.zig` now pre-scans for `--help` / `-h` and short-circuits before validation. Added regression tests in `phase2/zcc/src/cli_test.zig` for `--help -Wall` and `--unknown-thing --help`.
- B2: `.github/workflows/phase2-zig.yml` now hardcodes Zig 0.16.0 SHA256 values from `download/index.json` and verifies `/tmp/zig.tar.xz` before extraction with `sha256sum` or `shasum`.
- B3: `scripts/run-in-docker-phase2.sh` now checks `zig version == 0.16.0`; missing or mismatched versions reinstall from the pinned stable tarball.
- B4: `phase2/zcc/README.md` now uses `tests/smoke/hello.c` in the `-E` example.
- B5: `phase2/zcc/src/cli_test.zig` now includes `test_cli_handles_multiple_input_files` for `foo.c`, `bar.c`, `baz.c`.

Local verification excerpt:
```text
zig build                                            PASS
zig build test --summary all                         PASS (8/8 tests)
zig build smoke --summary all                        PASS
zig build fmt --summary all                          PASS
zig build lint --summary all                         PASS
./zig-out/bin/zcc --help                             PASS
./zig-out/bin/zcc --help -Wall                       PASS (prints usage, exit 0)
zig build run -- -E tests/smoke/hello.c              PASS
IMAGE=gcc:14 bash scripts/run-in-docker-phase2.sh zcc clean test smoke  PASS (retry after transient curl EOF)
```

Line-level Cubic findings were replied to directly with the fixed SHA.

{[zig-engineer]}

**Comment by code-yeongyu** [comment]
## `{[ci-engineer]} {[reviewer-quality]}` review — PR #35 Wave 0 at `984e9aaab3cde0b4ce38dab238aa297503c81caf`

**Verdict: APPROVE**

All 5 B-fixes (B1-B5) addressed. All 13 acceptance gate criteria met. Zero unresolved cubic findings at this head SHA. SHA256 supply-chain hashes verified against authoritative ziglang.org index.

### Verification protocol applied

Per the post-retraction protocol I committed to in [PR #34 issuecomment-4326971822](https://github.com/code-yeongyu/c11-compiler-zig-omo/pull/34#issuecomment-4326971822), all Zig 0.16 API claims verified against `/opt/homebrew/Cellar/zig/0.16.0_1/lib/zig/std/`. SHA256 hashes verified against `https://ziglang.org/download/index.json`.

### B-fix verification (all addressing my prior 17389dc RC)

| # | Fix | File:line at 984e9aa | Verification |
|---|---|---|---|
| B1 | `--help` short-circuits validation | `cli.zig:54-59` | Pre-pass loop scans args; `--help`/`-h` sets `config.help = true; return config;` BEFORE the validation while-loop. Verified. |
| B1 | Regression tests | `cli_test.zig:101-129` | `test_cli_help_short_circuits_unknown_options` (--help -Wall) + `test_cli_help_short_circuits_when_followed_by_unknown_flag` (--unknown-thing --help) — both confirm parse succeeds with `help == true`. Both use given/when/then. |
| B2 | SHA256 verification of Zig tarball | `.github/workflows/phase2-zig.yml:64-77` | 4 platform hashes hardcoded; `sha256sum -c` (Linux) / `shasum -a 256 -c` (macOS) gates extraction. **All 4 hashes match upstream**: aarch64-macos `b23d70de…`, x86_64-macos `0387557e…`, x86_64-linux `70e49664…`, aarch64-linux `ea4b09bf…` — verified via `curl ziglang.org/download/index.json` (timestamp:1776173999 from minisig). |
| B3 | Docker version pin | `scripts/run-in-docker-phase2.sh:20,55` | `ZIG_VERSION="0.16.0"` constant; container check `if [ "$installed_version" != "$ZIG_VERSION" ]; then install_zig; fi` re-installs if mismatch. Idempotent. |
| B4 | README example path | `phase2/zcc/README.md:16` | `zig build run -- -E tests/smoke/hello.c` — fixture exists at `phase2/zcc/tests/smoke/hello.c` (verified `git ls-tree`). |
| B5 | Multi-input test | `cli_test.zig:83-99` | `test "test_cli_handles_multiple_input_files"` parses `foo.c bar.c baz.c`, asserts `inputs.len == 3` + each filename. Goes BEYOND task #2 spec (required `len == 2`). |

### Acceptance gate (13/13 ✅)

| # | Gate | Status | Evidence |
|---|---|---|---|
| 1 | build.zig at `phase2/zcc/build.zig` with test/smoke/fmt/lint/run targets | ✅ | `build.zig:23,29,37,47,65` |
| 2 | Zig 0.16 idioms (`addExecutable(.{.root_module})`, `b.createModule`) | ✅ | `build.zig:7-14, 50-60` |
| 3 | `-Doptimize=Debug` not `-Drelease-safe` | ✅ | workflow doesn't hardcode obsolete flag |
| 4 | CI guard `phase2/zcc/build.zig`, `working-directory: phase2/zcc` | ✅ | `phase2-zig.yml:29, 70-83` |
| 5 | gcc:14 docker matrix cell green | ✅ | CI status `phase2-zig / docker / gcc-14 / zcc` SUCCESS |
| 6 | `scripts/run-in-docker-phase2.sh` exists, phase1 helper untouched | ✅ | new file; `scripts/run-in-docker.sh` unchanged |
| 7 | `build.zig.zon` zero deps | ✅ | file does not exist (zero-deps trivially satisfied via `@import("std")` only) |
| 8 | Tests in given/when/then, no mocks | ✅ | all 8 tests use given/when/then comments; no mocks |
| 9 | All 6 required tests from task #2 spec present | ✅ | help, compile flags, preprocessor+sanitizer, missing -o, unknown flag, multiple inputs (B5) — plus 2 bonus help-short-circuit regression tests |
| 10 | `zig fmt --check` passes | ✅ | CI step green |
| 11 | Sub-issue #16 success criteria 1-5 met | ✅ | (1) builds on macOS+docker, (2) `--help` exits 0, (3) `-E hello.c` runs preprocessor stub, (4) zero deps, (5) `zig build test` passes |
| 12 | All CI cells green at exact head SHA | ✅ | 7 SUCCESS (phase-2 guard, phase-3 guard, zcc ubuntu, zcc macos, docker gcc-14, GitGuardian, cubic) + 2 SKIPPED (zcc-vs-gcc, cross-compile — reserved for future waves) + 0 FAILURE |
| 13 | cubic SUCCESS + zero unresolved findings | ✅ | cubic check SUCCESS at 12:40:56Z; no new "issues found" review at 984e9aa; prior 17389dc findings (cli.zig:64, README.md:16) addressed by B1+B4 |

### Code quality review (DUAL-HAT scope)

**`{[reviewer-quality]}` lens** (cli/main/test code):
- ✅ No AI slop; clean idiomatic Zig 0.16 (`std.ArrayList .empty`, `std.heap.DebugAllocator` not present here but `std.Io.File.Writer`, `std.process.Init`)
- ✅ Strongly-typed `CliConfig` with sensible defaults
- ✅ Error union signature `(std.mem.Allocator.Error || ParseError)!CliConfig`
- ✅ Memory hygiene: `errdefer config.deinit(allocator)` + `defer sanitizers.deinit(allocator)` + `defer inputs.deinit(allocator)`
- ✅ Help short-circuit returns early — no leak (no allocations made yet)
- ✅ All tests use `std.testing.allocator` (catches leaks)
- ✅ Naming clear, no dead code, no over-engineering

**`{[ci-engineer]}` lens** (workflow + docker script):
- ✅ Concurrency block with `cancel-in-progress: true` (Phase 1 PR #15 pattern preserved)
- ✅ Manual zig install bypasses `mlugg/setup-zig` mirror outage; idempotent
- ✅ SHA256 supply-chain verification before extraction (defense against mirror tampering)
- ✅ Docker script idempotent: re-installs Zig only if version mismatch
- ✅ phase2 docker wrapper isolated from phase1 wrapper (no regression risk to Phase 1 CI)
- ✅ All 5 contract targets called in CI (fmt, build, test, smoke, lint)

### Prior P3 observations from 17389dc verdict (deferred to Wave 1 by lead — non-blocking)

- **O1**: `build.zig` lint uses `build-exe -fno-emit-bin` (functionally stricter than `ast-check`) — playbook deviation deferred per lead's Wave 1 follow-up
- **O2**: CI matrix lacks Debug + ReleaseFast variants per discipline doc — deferred per lead's Wave 1 follow-up

### Verdict

`reviewer-quality: APPROVE at 984e9aaab3cde0b4ce38dab238aa297503c81caf`

Dual-hat signed since this iteration touches build.zig contract surface + CI workflow + docker wrapper. Clean to merge as soon as the other 3 verdicts (verifier-deep, reviewer-ultrabrain, qa-tester) post.

— `{[ci-engineer]} {[reviewer-quality]}`

**Comment by code-yeongyu** [comment]
## QA report — {[qa-tester]}

**Head:** `984e9aaab3cde0b4ce38dab238aa297503c81caf`
**Date:** 2026-04-27
**Host:** macOS (darwin)

### Phase 2 zcc matrix (macOS native)

| Target | Result |
|---|---|
| `zig build` | PASS |
| `zig build test` | PASS — 8/8 tests passed |
| `zig build smoke` | PASS — token dump of hello.c |
| `zig build fmt` | PASS |
| `zig build lint` | PASS |
| `./zig-out/bin/zcc --help` | PASS — exit 0, usage printed |
| `./zig-out/bin/zcc --help -Wall` | PASS — exit 0 (B1 short-circuit fix verified) |
| `./zig-out/bin/zcc -E tests/smoke/hello.c` | PASS — tokens printed |

### Phase 1 regression

| Subproject | Result |
|---|---|
| c11-ref (make all + test) | PASS — all 23 tests OK |
| http2 (make all + test + smoke) | PASS — kqueue backend |
| doom (make all) | PASS — build completed |

### CI status at 984e9aa

| Check | Status |
|---|---|
| cubic · AI code reviewer | PASS (2m41s) |
| zcc build & test (macos-latest) | PASS (38s) |
| zcc build & test (ubuntu-latest) | PASS (23s) |
| phase2-zig / docker / gcc-14 / zcc | PASS (40s) |
| GitGuardian Security Checks | PASS (1s) |
| phase-2 guard | PASS (3s) |
| phase-3 guard | PASS (4s) |
| mergeStateStatus | CLEAN |
| mergeable | MERGEABLE |

### Blockers addressed verification

- **B1** (help short-circuit): `zcc --help -Wall` now exits 0 — verified locally ✓
- **B2** (CI SHA256): workflow updated — verified via CI PASS ✓
- **B3** (docker zig version pin): script updated — CI docker job PASS ✓
- **B4** (README path): verified `tests/smoke/hello.c` exists and works ✓
- **B5** (multi-input test): 8/8 tests passed (up from 5/5) — new tests confirmed ✓

### Verdict

**qa-tester: PASS at 984e9aaab3cde0b4ce38dab238aa297503c81caf** — all contract targets pass, 8/8 tests pass, CLI behaves correctly, Phase 1 baseline clean, CI fully green.

**Comment by code-yeongyu** [comment]
verifier-deep: PASS at 984e9aaab3cde0b4ce38dab238aa297503c81caf

Hands-on verification completed at exact head `984e9aaab3cde0b4ce38dab238aa297503c81caf`.

Verified:
- Exact worktree HEAD matches `984e9aaab3cde0b4ce38dab238aa297503c81caf`.
- Native contract passed: `zig build`, `zig build test --summary all` (8/8 tests), `zig build smoke --summary all`, `zig build fmt --summary all`, `zig build lint --summary all`, `./zig-out/bin/zcc --help`.
- B1 fixed: `./zig-out/bin/zcc --help -Wall` prints usage and exits 0; `-h` is supported; regression tests present.
- B2 fixed: workflow verifies Zig 0.16.0 tarball SHA256 before extraction with platform hash map.
- B3 fixed: docker runner enforces `zig version == 0.16.0`; mismatch probe with fake `zig 0.15.0` forced reinstall and yielded `/opt/zig/zig version` 0.16.0.
- B4 fixed: README uses `tests/smoke/hello.c`.
- B5 fixed: multiple-input-file regression test present.
- Docker gcc:14 passed: `IMAGE=gcc:14 bash scripts/run-in-docker-phase2.sh zcc clean test smoke`.
- CI rollup at exact head is green/expected-skipped only; cubic SUCCESS; mergeStateStatus CLEAN/MERGEABLE.

{[verifier-deep]}

**Comment by code-yeongyu** [comment]
[reviewer-ultrabrain]

reviewer-ultrabrain: APPROVE at 984e9aaab3cde0b4ce38dab238aa297503c81caf

Re-audit summary:
- B1 help short-circuit is structurally safe for Wave 0: `--help` / `-h` pre-pass returns a default help config before unknown-option validation, and regression tests cover unknown flags before/after help.
- B2 CI SHA256 values match Zig 0.16.0 `download/index.json` for aarch64-macos, x86_64-macos, x86_64-linux, and aarch64-linux.
- B3 Docker runner enforces `zig version == 0.16.0` by reinstalling into `/opt/zig` when absent or mismatched.
- B4 README path now matches the actual smoke fixture: `tests/smoke/hello.c`.
- B5 multi-input parser path is covered by `test_cli_handles_multiple_input_files`; manual `zcc -E` with two inputs also emits both token streams.

Local verification on Zig 0.16.0:
- `zig build fmt`
- `zig build`
- `zig build test --summary all` (8/8 tests)
- `zig build smoke --summary all`
- `zig build lint --summary all`
- `./zig-out/bin/zcc --unknown-thing --help`
- `./zig-out/bin/zcc -h -Wall`
- `./zig-out/bin/zcc -E tests/smoke/hello.c tests/smoke/hello.c`

CI gate observed clean at this head: cubic SUCCESS, GitGuardian SUCCESS, ubuntu/macOS zcc SUCCESS, phase2 docker gcc:14 SUCCESS, mergeStateStatus CLEAN.

Note: Zig LSP diagnostics could not run because `zls` is configured but not installed; compiler/build verification passed. YAML diagnostics reported no findings for the workflow.
